lib/Dialect/HW/InnerSymbolTable.cpp

```diff
@@ -78,33 +147,124 @@ LogicalResult InnerSymbolTable::walkSymbols(Operation *op,
+    if (auto mod = dyn_cast<PortList>(op)) {
+      for (auto [i, port] : llvm::enumerate(mod.getPortList())) {
+        if (auto symAttr = port.getSym()) {
+          if (failed(walkSyms(symAttr, InnerSymTarget(i, op, 0))))
+            return failure();
+        }
+      }
+    }
+
+    return success();
+  };
+
+  // First, check the operation itself for ports. This is a separate step
+  // since we want ports to be listed in the InnerSymbolTable of the operation
+  // itself, and not in the parent InnerSymbolTable.
+  if (failed(walkPorts(op)))
+    return failure();
+
+  // Perform the walk. We do not use Operation::walk here, since recursing
+  // into regions must go through InnerSymbolTable::walkSymbolsInner to properly
+  // track nested symbol tables.
+  for (auto &region : op->getRegions()) {
+    for (auto &block : region.getBlocks()) {
+      for (auto &innerOp : block) {
+        // Check for nested symbol table.
+        if (innerOp.hasTrait<OpTrait::InnerSymbolTable>()) {
+          if (failed(walkTable(InnerSymbolTable::getName(&innerOp), &innerOp)))
+            return failure();
+        }
+
+        // Check for InnerSymbolOpInterface
+        if (auto symOp = dyn_cast<InnerSymbolOpInterface>(&innerOp))
+          if (auto symAttr = symOp.getInnerSymAttr())
+            if (failed(walkSyms(symAttr, InnerSymTarget(symOp, 0))))
+              return failure();
+
+        // Check for port-attached symbols.
+        if (failed(walkPorts(&innerOp)))
+          return failure();
+
+        // Recurse into the operation.
+        if (failed(walkSymbolsInner(&innerOp, symCallback, symTblCallback,
+                                    currentIST)))
+          return failure();
+      }
+    }
+  }
+
+  return success();
+}
+
+LogicalResult InnerSymbolTable::walkSymbols(
+    Operation *op, InnerSymCallbackFn symCallback,
+    std::optional<InnerSymTableCallbackFn> symTblCallback) {
+  return walkSymbolsInner(op, symCallback, symTblCallback,
+                          /*currentSymbolTable*/ nullptr);
+}
+
+InnerSymTarget InnerSymbolTable::lookup(hw::InnerRefAttr path) const {
+  return lookup(path.getPath());
+}
+
+InnerSymTarget InnerSymbolTable::lookup(llvm::ArrayRef<StringAttr> path) const {
+  assert(!path.empty() && "Expected non-empty path");
+  assert(llvm::all_of(path, [](auto &attr) { return attr; }) &&
+         "Expected non-null path elements");
+
+  if (path.size() == 1) {
+    // This was the last element in the path, lookup the symbol.
+    return lookup(path.front());
+  }
+
+  // Else, we have a path with more than one element; this implies that this
+  // target should be a nested symbol table.
+  auto nestedSymTbl = nestedSymbolTables.find(path.front());
+  if (nestedSymTbl == nestedSymbolTables.end())
+    return {};
+
+  // Perform the internal lookup into the nested symbol table.
+  return nestedSymTbl->second->lookup(path.drop_front());
+}
+
+void InnerSymbolTable::dump(llvm::raw_ostream &os, int indent) const {
```

Comment:
`dump()` usually doesn't take an ostream/indent, so you can just `dump()`, say, from a debugger.

Usually it uses the `dbgs()` and calls some other method, e.g., `print`.

There are also macros for these so they play well with the larger LLVM project (drop in full release build, optionally enable them regardless, always have in debug builds), for example:
https://github.com/llvm/llvm-project/blob/26060de06378375f784b35df80ec6d43425e5908/llvm/lib/MC/MCInst.cpp#L65